### PR TITLE
Add support for Emacs-25.0.94 pre-test

### PR DESCRIPTION
**Pull request by phillord**
As discussed in #69.

I've had to refactor some parts -- particularly I've removed the "z" from "tar xvfz". My tar just copes with this (and does xv files also). Don't know about the mac tar though.

### CODE REVIEW

File: recipes/emacs-25.0.94.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+recipe 'emacs-25.0.94' do
+  tar_xv 'http://alpha.gnu.org/gnu/emacs/pretest/emacs-25.0.94.tar.xz'
```
Comment (rejeep) [inline recipes/emacs-25.0.94.rb]: Why are you calling the method `tar_xv` when the extension is `.xz`?

**Comment by rejeep** [comment]
1. This only adds support for a compiled version and not a Travis version.
2. I'm not sure about calling it `emacs-25.0.94`. What happens when there's another prerelease? Do we create another recipe then? If we call it `emacs-25-prerelease`, we only need to update the path to the tar-ball and build a new travis version.

I also wonder what the real purpose of this is, specially as you didn't add a Travis version?

File: recipes/emacs-25.0.94.rb (hunk 1 above)
Comment (phillord) [inline recipes/emacs-25.0.94.rb]: xv --- showing my age, used to use this command all the time.

**Comment by phillord** [comment]
Addressed most of these -- not sure how you work out the URL for the travis version.

Agree about the "prerelease" tag. Decided one way, then changed my mind, but decide that this is probably best.

**Comment by rejeep** [comment]
Well, it's a little bit of work every time a new Travis version is added, that's why I'm a big sceptical about it. At least I have the procedure documented: https://github.com/rejeep/evm#adding-travis-binary

**Comment by phillord** [comment]
I don't mind making the binaries, in so far as I have permission.

I'm not sure that I understand the instructions for generating the binary. What is the purpose of vagrant here? Can the binaries not just be produced in a host linux box? Also, why is the `make bootstrap`  necessary -- the `make install` should do this anyway but `make bootstrap` will force recompilation of the elc files (which is pretty slow).

**Comment by rejeep** [comment]
> What is the purpose of vagrant here? Can the binaries not just be produced in a host linux box?

As long as they are compiled on the same architecture as on Travis I guess it's ok. But since Travis provides a very simple way of setting up an exact copy of their environment, why not use it?

> Also, why is the make bootstrap necessary

I usually compile from a new tar-ball every time I compile Emacs and then it doesn't matter I guess.

I may be wrong, but doesn't the `INSTALL` say bootstrap should be used?

**Comment by phillord** [comment]
`make bootstrap` is a quite a bit slower because the src tar ball normally contains the elc files. It's good when bisecting but otherwise, not really needed.

I build a binary tonight and put it somewhere useful!

**Comment by rejeep** [comment]
> make bootstrap is a quite a bit slower because the src tar ball normally contains the elc files. It's good when bisecting but otherwise, not really needed.

Then I guess recipes where Emacs is compiled from scratch could be updated, for example: https://github.com/rejeep/evm/blob/master/recipes/emacs-24.5.rb#L16-L18

> I build a binary tonight and put it somewhere useful!

I store all pre compiled binaries under each release, see: https://github.com/rejeep/evm/releases

**Comment by phillord** [comment]
Think that's it!

**Comment by phillord** [comment]
Just pinging. Think this is complete now. I have added a binary on http://www.russet.org.uk/scratch.

File: lib/evm/builder.rb [hunk 2]
```diff
@@ -23,8 +23,8 @@ def git(url)
         end
       end
 
-      def tar_gz(url)
-        tar_file_path = File.join(builds_path, @name + '.tar.gz')
+      def tar_packaged(url,extension)
```
Comment (rejeep) [inline lib/evm/builder.rb]: Can you please make this private so that it's not available in the DSL.

File: lib/evm/builder.rb [hunk 3]
```diff
@@ -35,6 +35,14 @@ def tar_gz(url)
         tar_file.extract(builds_path, @name)
       end
 
+      def tar_xz(url)
+        tar_packaged(url,'xz')
+      end
+
+      def tar_gz(url)
+        tar_packaged(url,'gz')
```
Comment (rejeep) [inline lib/evm/builder.rb]: Some spacing after `,` here and above.

File: recipes/emacs-25-pre-travis.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+recipe 'emacs-25-pre-travis' do
+  tar_gz 'https://github.com/rejeep/evm/releases/download/v0.6.1/emacs-25-pre-travis.tar.gz'
```
Comment (rejeep) [inline recipes/emacs-25-pre-travis.rb]: Change this to `https://github.com/rejeep/evm/releases/download/v0.8.0/emacs-25-pre-travis.tar.gz `. I will release `v0.8.0` with this change and attach the binary to that release.

File: recipes/emacs-25-pre.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  linux do
+    option '--prefix', installation_path
+    option '--without-gif'
+  end
+
+  install do
+    configure
+    make 'bootstrap'
```
Comment (rejeep) [inline recipes/emacs-25-pre.rb]: If you are saying we don't need bootstrap, then perhaps we should skip it?

**Comment by rejeep** [comment]
@phillord Sorry for the delay, I have very little time left over with a six month old... 😕 

I made a few comments, please adress those and I will try to get this in production as soon as possible.

**Comment by phillord** [comment]
@rejeep No worries about the delay. Having a six month at home is cause for celebration.

I've made all the changes. I have checked -- make bootstrap does not seem necessary (and the compile is far quicker, since no byte compilation is needed).

**Comment by rejeep** [comment]
@phillord All done now, wanna try it out please!?

**Comment by phillord** [comment]
Alas, no, it is not working.

Something strange is going on. I am getting three copies of Emacs

https://travis-ci.org/phillord/assess/builds/136210347

which does `find /tmp -type d`

There are two copies of Emacs-24.3 installed by cask, one at `/tmp/emacs-24.3` and one at `/tmp/tmp/emacs-24.3`. And Emacs 25 has been installed at `/tmp/tmp/emacs-25-pre`. Oh dear.

**Comment by rejeep** [comment]
I guess you are using https://gist.github.com/rejeep/ebcd57c3af83b049833b#file-travis-evm-cask-sh-L22?

> one at /tmp/tmp/emacs-24.3

This is where the tar-ball in unzipped and compiled.

> one at /tmp/emacs-24.3

That's the real installation path from the script above (`--prefix`)

> And Emacs 25 has been installed at /tmp/tmp/emacs-25-pre

There should also be a `/tmp/emacs-25-pre`

**Comment by phillord** [comment]
Okay, checked. So, it appears that you were entirely correct -- the binaries are not working on travis. So, the whole vagrant step is necessary. 

Oh dear, my bad. I will try and get this bit working over the next few days. Apologies for time waste.

**Comment by rejeep** [comment]
No worries, let me know when you are done.

**Comment by phillord** [comment]
New binary is here:

http://www.russet.org.uk/scratch/emacs-25-pre-travis.tgz

It's actually 25.0.95 as there has been a new pre-test in the meantime. I've worked out how to test it now, and you can see an Emacs hello world here:

https://travis-ci.org/phillord/evm-tester
 
Again, apologies for making you do this twice!

**Comment by rejeep** [comment]
Updated, can you please try again!?

> Again, apologies for making you do this twice!

No worries, just glad I don't have to do the compiling myself... 😄

**Comment by phillord** [comment]
Ack --- I the used wrong extension on the tarball -- .tgz not .tar.gz. So, it 404s at the moment.

**Comment by rejeep** [comment]
Would it work to simply rename it?

**Comment by rejeep** [comment]
I renamed it

**Comment by rejeep** [comment]
@phillord I added you as a collaborator to the repo

**Comment by phillord** [comment]
Should do, yes. It's just the name.

**Comment by phillord** [comment]
All sorted -- it's working well now. And accepted collaborator.

Once Emacs 25 is released, I'll think about doing a build from master branch if you are happy.

**Comment by syohex** [comment]
Emacs 25.1 is released now.

**Comment by phillord** [comment]
Yes. I'll make these binaries over the next couple of days.

**Comment by rejeep** [comment]
See https://github.com/rejeep/evm/pull/73

**Pull request merged**